File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode last = null;
        ListNode cur = head;
        while (cur != null) {
            ListNode next = cur.next;
            cur.next = last;
            last = cur;
            cur = next;
        }
        return last;
    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.kt
/**
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun removeNthFromEnd(head: ListNode?, n: Int): ListNode? {
        var count = 0
        var cur: ListNode? = head
        while (cur != null) {
            cur = cur?.next
            count++
        }
        if (n == count) {
            return head?.next
        }
        cur = head
        count = count - n - 1
        for (i in 1..count) {
            cur = cur?.next
        }
        cur?.next = cur?.next?.next
        return head
    }
}
